ledger/onlineaccountscache.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package ledger
+
+import (
+	"github.com/algorand/go-algorand/data/basics"
+)
+
+type onlineAccountsCache struct {
+	accounts map[basics.Address]*persistedOnlineAccountDataList
+}
+
+// init initializes the onlineAccountsCache for use.
+// thread locking semantics : write lock
+func (o *onlineAccountsCache) init() {
+	o.accounts = make(map[basics.Address]*persistedOnlineAccountDataList)
+}
+
+// read the persistedAccountData object that the cache has for the given address.
+// thread locking semantics : read lock
+func (o *onlineAccountsCache) read(addr basics.Address, rnd basics.Round) (data persistedOnlineAccountData, has bool) {
+	if list := o.accounts[addr]; list != nil {
+		node := list.back()
+		if node.Value.round > rnd {
+			return persistedOnlineAccountData{}, false
+		}
+		for node.prev != &list.root {
+			node = node.prev
+			// only need one entry that is targetRound or older
+			if node.Value.round > rnd {
+				return *node.next.Value, true
+			}
+		}
```

Comment:
this is potentially very inefficient for accounts changing their state rapidly (ie balance change).